src/System.Linq/src/System/Linq/Reverse.cs

```diff
@@ -39,19 +38,25 @@ public override Iterator<TSource> Clone()
 
             public override bool MoveNext()
             {
+                if (_state - 2 <= -2)
```

Comment:
@wangyanjun That is going to be different if `_state` is `int.MinValue` or `int.MinValue + 1`. We need to handle those numbers.